### ISSUE TITLE
OpenSSL CMS_verify 3.0.0 Error

### ISSUE THREAD

**Issue by hafedh-trimeche** [openssl/openssl#16627]
CMS_verify 3.0.0 returns this error:
`error:17000064:CMS routines::certificate verify error
`
The 1.1.1l version executed without error.

Flags used:   CMS_NO_SIGNER_CERT_VERIFY or CMS_BINARY or CMS_NOSMIMECAP or CMS_CADES;

Thanks.

**Comment by paulidale** [comment]
Are you loading the [legacy provider](https://www.openssl.org/docs/man3.0/man7/migration_guide.html#Legacy-Algorithms)?  It can be loaded via configuration, function call or the `-provider` argument to our commands.

If not, I suspect more will be needed to figure this out.  At the minimum the certificate.

**Comment by hafedh-trimeche** [comment]
Please find attached the binary CMS with all embedded certificates.
[CMS.zip](https://github.com/openssl/openssl/files/7194496/CMS.zip)

**Comment by paulidale** [comment]
Are you loading the legacy provider?

**Comment by hafedh-trimeche** [comment]
No,

How can it be done programmatically ?

**Comment by paulidale** [comment]
Yes:
```c
OSSL_PROVIDER_load(NULL, "legacy");
OSSL_PROVIDER_load(NULL, "default");
```

**Comment by hafedh-trimeche** [comment]
Please note that there is no **OSSL_PROVIDER_free** function located in **libcrypto-3-x64.dll** download from https://slproweb.com/products/Win32OpenSSL.html.
problem not solved when Provider set to legacy.

Please find the Pascal code used (with wrapped functions):
```
function CMSVerify(const Blob:TBytes;DetachedContent:TBytes;const CAdES:Boolean):TCMSInfo;
var
  MemIO           : PBIO;
  CMS             : PCMS_ContentInfo;
  InfoStack       : PSTACK_OF_CMS_SignerInfo;
  SignerInfo      : PCMS_SignerInfo;
  SignerInfoCount ,
  iSignerInfo     : Integer;
  CMSCerts        : PSTACK_OF_X509;
  Cert            : PX509;
  CertIdx         : Integer;
  CertsCount      ,
  iCerts          : Integer;
  CertInfo        : TCertInfo;
  Flags           : Integer;
begin
  Initrecord(Result,SizeOf(Result));
  if Blob=nil then Exit;
  Result.Content := Blob;
  MemIO          := BIONew(Blob);
  CMS            := d2i_CMS_bio(MemIO,nil);
  if not Assigned(CMS) then
  begin
    BIOReset(MemIO);
    CMS := SMIME_read_CMS(MemIO,nil);
  end;
  Free_BIO(MemIO);
  if not Assigned(CMS) then Exit;
  InfoStack       := CMS_get0_SignerInfos(CMS);
  SignerInfoCount := OPENSSL_sk_num(InfoStack);
  CMSCerts        := nil;
  if SignerInfoCount>0 then
  begin
    Result.eContentType := OBJ_obj2nid(CMS.signedData.encapContentInfo.eContentType);
    CMSCerts            := CMS_get1_certs(CMS);
    CertsCount          := OPENSSL_sk_num(CMSCerts);
    for iCerts:=CertsCount-1 downto 0 do
    begin
      Cert := OPENSSL_sk_value(CMSCerts,iCerts);
      X509Parse(X509ToPEM(Cert),CertInfo);
      if CertInfo.CertType in [CrtCA,CrtRoot] then
      begin
        CertIdx := Length(Result.Certs);
        SetLength(Result.Certs,CertIdx+1);
        Result.Certs[CertIdx] := CertInfo;
      end;
      if Result.Signer.CommonName='' then
      begin
        for iSignerInfo:=0 to SignerInfoCount-1 do
        begin
          SignerInfo := OPENSSL_sk_value(InfoStack,iSignerInfo);
          if CMS_SignerInfo_cert_cmp(SignerInfo,Cert)=0 then
          begin
            CertIdx := Length(Result.Certs);
            SetLength(Result.Certs,CertIdx+1);
            Result.Certs[CertIdx] := CertInfo;
            if Result.Signer.CommonName='' then Result.Signer := CertInfo;
            Break;
          end;
        end;
      end;
    end;
  end;
  Flags := CMS_NO_SIGNER_CERT_VERIFY or CMS_BINARY or CMS_NOSMIMECAP;
  if CAdES then Flags := Flags or CMS_CADES;
  if Result.eContentType=NID_pkcs7_data then
  begin
    if Assigned(DetachedContent) then MemIO := BIONew(DetachedContent)
                                 else MemIO := nil;
  end
  else
  begin
    MemIO := nil;
    Flags := Flags or CMS_NO_CONTENT_VERIFY;
  end;
  if CMS_verify(cms,nil,nil,MemIO,nil,Flags)>0 then Result.Verified := True else
  begin
    Result.ErrorMessage := SSLErrorMessage('CMS_verify',Result.ErrorCode);
    {$IFDEF DEBUG}
    if IsDevMachine then
    begin
      if Assigned(DetachedContent) then
      begin
        DebugWrite(ApplicationLogPath+'CMS-Detached.bin',Blob);
        DebugWrite(ApplicationLogPath+'Detached.bin',DetachedContent);
      end
      else DebugWrite(ApplicationLogPath+'CMS.bin',Blob);
    end;
    {$ENDIF}
  end;
  Free_sk_pop(CMSCerts);
  Free_BIO(MemIO);
  Free_CMS_ContentInfo(CMS);
end;

```

**Comment by hafedh-trimeche** [comment]
Just figured out that **OSSL_PROVIDER_unload** should be used instead of **OSSL_PROVIDER_free**.

**Comment by paulidale** [comment]
Yeah, it's unload not free.  It was late :disappointed:

**Comment by paulidale** [comment]
You will need _both_ the legacy and the default provider loaded.

**Comment by hafedh-trimeche** [comment]
Both providers loaded but the same error is generated.

**Comment by hafedh-trimeche** [comment]
```
OSSL_PROVIDER_load(NULL, "legacy");
OSSL_PROVIDER_load(NULL, "default");
```
Removed from initialization.

**CMS_CADES** also removed from **CMS_verify** flags which returned successful verification. 

Why CMS_CADES caused the error? 1.1.1l version doesn't.

**Comment by paulidale** [comment]
CMS_CADES doesn't exist in 1.1.1.

The documentation has this to say about CMS_CADES:
> If **CMS_CADES** is set, each signer certificate is checked against the ESS signingCertificate or ESS signingCertificateV2 extension that is required in the signed attributes of the signature.

**Comment by hafedh-trimeche** [comment]
Please find attached a report from https://signatures-conformance-checker.etsi.org/ which states that the signature **PAdES [ETSI.CAdES.detached]** signed by **seller.strong-data.com** is conformant with ETSI EN 319 142-2 v1.1.1 Additional PAdES signatures profiles.

The attributes are set to **essSigningCertificateV2**.

But CMS_Verify returns: error:17000064:CMS routines::certificate verify error 
[CMS_Verify.zip](https://github.com/openssl/openssl/files/7209827/CMS_Verify.zip)

Would more error details be provided by the OpenSSL API (Signer, Certificate SN, ...)?

**Comment by nhorman** [comment]
Marking as inactive, will be closed at the end of 3.4 dev barring further input.
